### DISCUSSION TITLE
What is the "correct" way to configure a package?

### DISCUSSION THREAD

**Discussion by davystrong** [typst/typst#4787] [Q&A] [answered]
I was recently using the [gentle-clues](https://typst.app/universe/package/gentle-clues/) package and noticed that it didn't play well with one of my bare show rules. When I looked into it, I realised that the clue was appearing as an opaque context. I read up on [context](https://typst.app/docs/reference/context/) and it does mention this, I just had never noticed.

I looked into the [source code](https://github.com/jomaway/typst-gentle-clues/blob/main/lib/clues.typ) for clue and it appears the only reason the author uses context is to allow default parameters to be set at import time:

```
#import "@preview/gentle-clues:0.9.0": *
#show: gentle-clues.with(...)

// Then later
#clue[...]
```

This feels like a sledgehammer to crack a nut, especially when the default settings are likely useful to most people anyway, and it results in the output being hidden in an opaque context block.

My question is, what is the "correct" way to write a package and allow the user to set default values once for all function calls? Is this the best way? If so, is there any way to see inside the context block, i.e. from another context block or something?

I realise this is effectively implementing `set` statements on custom functions before that is officially possible. However, I read somewhere (sorry, can't find where) that accessing the value set by a `set` would still require context, so that wouldn't solve the problem.

**Comment by Mc-Zen** [comment, marked as answer]
Hi @davystrong 

This is actually a very good question. I think everyone would agree that using state and context is in principle _not_ the ideal way of doing that. However, it is currently more or less the _only_ way 😅. 

In the (hopefully rather near) future, it will be possible to write _user-defined_ types. For these, one can use `#set` and `#show` rules making configuration much simpler and more performant. 

Another way of achieving configuration is using _factory functions_. This is also used in some packages. As an example, you could have a `setup()` function that returns a pre-configured version of  `clue` (sticking to your example here). 

```typ
#let clue = setup-clue(...) // let's imagine gentle-clues did it like that
// This also works with multiple functions
#let (clue, tip)= setup-clue(...)
```

Or even just (this is generally always possible)

```typ
#import "@preview/gentle-clues:0.9.0": *
#let clue = clue.with(...)
```

The drawback is that the `clue()` function is configured this way as bound to the variable. If you import the package in question in a second file (`#include`d in the main file) you need to configure again. Or better: add all imports to a `template.typ` or similar and import that in every file of your document. 


---

Even in the case of (future) _user-defined_ types, I think it might not be possible to "look" into the content. What do you want to achieve here that is hindered by the opaque context? Usually, it is not a good idea to inspect content too much but there might be a different way of going about it.

**Reply by davystrong** [reply]
Thanks @Mc-Zen for your reply! I had tried something like the factory functions method you mentioned with the template, however this didn't work as expected either:

```
// template.typ
#let innerfunc(a: 6) = {
  [#a]
}

#let template(document) = {
  let func = innerfunc.with(a: 4)
  document
}

// main.typ
#import "template.typ": *

#show: template

This is the output of innerfunc: #innerfunc()
This is the output of func: #func()
```

In this example, `func` can't be found ("unknown variable: func"), which is unexpected to me, though I may be doing something stupid.

Regarding why I want to look into context, that is probably the more pertinent question: it's a workaround for #86, so I guess it's not intended anyway. However, it does feel like it has the potential to be very annoying when used correctly too.

**Reply by Mc-Zen** [reply]
> In this example, func can't be found ("unknown variable: func"), which is unexpected to me, though I may be doing something stupid.

This one is easy luckily. Since you declare `func` in the scope of the function `template(document)`, it is not available in the global scope and the rule `#show: template` only applies rules (does not bring the definitions into scope). Here, you can just move the declaration of `func` out of `template()`. 

>Regarding why I want to look into context, that is probably the more pertinent question: it's a workaround for 86, so I guess it's not intended anyway. However, it does feel like it has the potential to be very annoying when used correctly too.

Ah I see. I'm afraid this won't be possible as of now.

**Comment by laurmaedje** [comment]
> I realise this is effectively implementing set statements on custom functions before that is officially possible. However, I read somewhere (sorry, can't find where) that accessing the value set by a set would still require context, so that wouldn't solve the problem.

Set and show on custom types will indeed still require context. There are various workarounds (like the ones for #86) that peek into content to do something, which breaks with context. But I think that is more a problem with these workarounds than with context. Contextual things are very foundational to Typst and will likely become more widespread with custom types and set rules rather than less.

See also: https://github.com/typst/typst/discussions/4745

**Reply by davystrong** [reply]
Ok, thanks for that! Yes, I accept the problem may simply be with the workarounds. And I don't understand how context works internally so I can't comment on how it could be changed to be less opaque (and assume it can't).
